Dynamic filtering for semi-joins: planner

lxynov opened the pull request:
Supersedes the planner part of #2190

sopel39 commented on presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestRemoveUnsupportedDynamicFilters.java:
```diff
@@ -319,6 +320,77 @@ public void testSpatialJoin()
     }
 
+    @Test
+    public void testSemiJoin()
```
keep this test cases separate

sopel39 commented on presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestRemoveUnsupportedDynamicFilters.java:
```diff
@@ -319,6 +320,77 @@ public void testSpatialJoin()
+                removeUnsupportedDynamicFilters(root),
+                semiJoin("ORDERS_OK", "LINEITEM_OK", "SEMIJOIN_OUTPUT", Optional.empty(), false,
+                        filter("ORDERS_OK > 0",
+                                tableScan("orders", ImmutableMap.of("ORDERS_OK", "orderkey"))),
```
put "orders" argument in newline

sopel39 commented on presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestRemoveUnsupportedDynamicFilters.java:
```diff
@@ -319,6 +320,77 @@ public void testSpatialJoin()
+        assertPlan(
+                removeUnsupportedDynamicFilters(root),
+                semiJoin("ORDERS_OK", "LINEITEM_OK", "SEMIJOIN_OUTPUT", Optional.empty(), false,
+                        filter("ORDERS_OK > 0",
```
use `io.prestosql.sql.planner.assertions.PlanMatchPattern#filter(io.prestosql.sql.tree.Expression, io.prestosql.sql.tree.Expression, io.prestosql.sql.planner.assertions.PlanMatchPattern)` version of filter to assert DF is really gone

sopel39 commented on presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestRemoveUnsupportedDynamicFilters.java:
```diff
@@ -319,6 +320,77 @@ public void testSpatialJoin()
+        assertPlan(
+                removeUnsupportedDynamicFilters(root),
+                semiJoin("ORDERS_OK", "LINEITEM_OK", "SEMIJOIN_OUTPUT", Optional.empty(), false,
+                        filter("ORDERS_OK > 0",
```
use `io.prestosql.sql.planner.assertions.PlanMatchPattern#filter(io.prestosql.sql.tree.Expression, io.prestosql.sql.tree.Expression, io.prestosql.sql.planner.assertions.PlanMatchPattern)`

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/sanity/DynamicFiltersChecker.java:
```diff
@@ -98,6 +100,25 @@ public void validate(PlanNode plan, Session session, Metadata metadata, TypeAnal
                 return ImmutableSet.copyOf(unmatched);
             }
 
+            @Override
```
please add a test to `TestDynamicFiltersChecker`

sopel39 reviewed (commented):
some comments, overall looks good

sopel39 commented:
please make sure tests pass

lxynov commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/PlanPrinter.java:
```diff
@@ -451,6 +451,7 @@ public Void visitSemiJoin(SemiJoinNode node, Void context)
             node.getDistributionType().ifPresent(distributionType -> nodeOutput.appendDetailsLine("Distribution: %s", distributionType));
             node.getSource().accept(this, context);
             node.getFilteringSource().accept(this, context);
+            node.getDynamicFilterId().ifPresent(dynamicFilterId -> nodeOutput.appendDetailsLine("dynamicFilterId: %s", dynamicFilterId));
```
```
        └─ SemiJoin[ss_sold_date_sk = d_date_sk][$hashvalue, $hashvalue_0]                                                                                                                                
           │   Layout: [ss_sold_date_sk:bigint, ss_promo_sk:bigint, $hashvalue:bigint, expr:boolean]                                                                                                      
           │   Estimates: {rows: 120527 (3.25MB), cpu: 15.92M, memory: 545B, network: 545B}                                                                                                               
           │   Distribution: REPLICATED                                                                                                                                                                   
           │   dynamicFilterId: df_439   
```

lxynov commented on presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:
```diff
@@ -326,6 +306,103 @@ public void testNonPushedDownJoinFilterRemoval()
+                                                anyTree(
+                                                        tableScan("lineitem", ImmutableMap.of("Y", "orderkey"))))))));
+
+        assertPlan("SELECT orderkey IN (SELECT orderkey FROM lineitem WHERE linenumber < 0) FROM orders",
```
it's a non-filtering semi join since it returns boolean values instead of filtering out rows?

lxynov commented on presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:
```diff
@@ -326,6 +306,103 @@ public void testNonPushedDownJoinFilterRemoval()
+                                                                anyTree(
+                                                                        filter("S1",
+                                                                                project(
+                                                                                        semiJoin("LINEITEM_OK", "ORDERS_OK", "S1", Optional.empty(), true,
```
yeah but I feel we can keep it as a comparison with the outer one, WDYT?

lxynov commented:
@sopel39 Thanks for the review! I've addressed the comments. All tests pass except a flaky one `TestPostgreSqlTypeMapping#testDecimalExceedingPrecisionMaxWithNonExceedingIntegerValues`.

sopel39 reviewed (approved):
small comments

lxynov commented on presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:
```diff
@@ -326,6 +306,177 @@ public void testNonPushedDownJoinFilterRemoval()
+                "SELECT * FROM orders WHERE orderkey IN (SELECT orderkey FROM lineitem WHERE linenumber % 4 = 0)",
+                anyTree(
+                        filter(
+                                "S",
```
agreed, this looks much nicer

lxynov commented on presto-main/src/test/java/io/prestosql/sql/planner/assertions/SemiJoinMatcher.java:
```diff
@@ -64,6 +76,26 @@ public MatchResult detailMatches(PlanNode node, StatsProvider stats, Session ses
+                    .stream()
+                    .flatMap(filterNode -> extractExpressions(filterNode).stream())
+                    .flatMap(expression -> extractDynamicFilters(expression).getDynamicConjuncts().stream())
+                    .anyMatch(descriptor -> descriptor.getId().equals(dynamicFilterId));
```
oh I missed that one while refactoring, thanks for the catch!

lxynov commented on presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:
```diff
@@ -326,6 +306,177 @@ public void testNonPushedDownJoinFilterRemoval()
+    {
+        // Dynamic filters are supported only after a table scan
+        assertPlan(
+                "WITH t AS (SELECT lineitem.partkey + 1000 partkey FROM lineitem WHERE lineitem.orderkey IN (SELECT orders.orderkey FROM orders)) " +
```
With `WITH t AS (SELECT lineitem.partkey + 1000 partkey FROM lineitem)`, the project will be merged with table scan but we don't have `TableScanAndProjectMatcher` to represent it. Would it be good to keep the inner semi-join here as it can also serve as a comparison to the outer semi-join? We can see clearly that the inner semi-join has dynamic filter but not the outer one.

sopel39 replied:
I think
```
SELECT partkey FROM (SELECT lineitem.partkey + 1000 partkey FROM lineitem) WHERE partkey IN (SELECT part.partkey FROM part)
```
should be sufficient.
It would be easier for reader to undestand the test intent and matched tree.

lxynov commented on presto-main/src/test/java/io/prestosql/sql/planner/AbstractPredicatePushdownTest.java:
```diff
@@ -122,11 +122,10 @@ public void testNonDeterministicPredicateDoesNotPropagateFromFilteringSideToSour
-                                                "LINE_ORDER_KEY", "orderkey"))),
+                        semiJoin("LINE_ORDER_KEY", "ORDERS_ORDER_KEY", "SEMI_JOIN_RESULT", enableDynamicFiltering,
+                                        anyTree(
+                                                tableScan("lineitem", ImmutableMap.of(
```
The diff here is misleading. This line is actually from a previous commit `9e066115 Push SemiJoin predicate inferred from filter side to source side`, so let's keep it as it is

lxynov commented:
@sopel39 Thanks for the review! PR updated.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -1184,7 +1184,7 @@ private PlanNode visitNonFilteringSemiJoin(SemiJoinNode node, RewriteContext<Exp
             PlanNode output = node;
             if (rewrittenSource != node.getSource() || rewrittenFilteringSource != node.getFilteringSource()) {
-                output = new SemiJoinNode(node.getId(), rewrittenSource, rewrittenFilteringSource, node.getSourceJoinSymbol(), node.getFilteringSourceJoinSymbol(), node.getSemiJoinOutput(), node.getSourceHashSymbol(), node.getFilteringSourceHashSymbol(), node.getDistributionType());
+                output = new SemiJoinNode(node.getId(), rewrittenSource, rewrittenFilteringSource, node.getSourceJoinSymbol(), node.getFilteringSourceJoinSymbol(), node.getSemiJoinOutput(), node.getSourceHashSymbol(), node.getFilteringSourceHashSymbol(), node.getDistributionType(), Optional.empty());
```
We could break that into multiple lines at this point

sopel39 commented on presto-main/src/test/java/io/prestosql/sql/planner/TestQuantifiedComparison.java:
```diff
@@ -35,12 +35,13 @@
-                                semiJoin("X", "Y", "S",
-                                        anyTree(tableScan("orders", ImmutableMap.of("X", "orderkey"))),
-                                        anyTree(values(ImmutableMap.of("Y", 0))))))));
+        assertPlan(query,
```
nit for future: this should be separate commit (it changes formatting)

sopel39 commented on presto-main/src/test/java/io/prestosql/sql/planner/assertions/SemiJoinMatcher.java:
```diff
@@ -64,6 +77,31 @@ public MatchResult detailMatches(PlanNode node, StatsProvider stats, Session ses
+                        .flatMap(expression -> extractDynamicFilters(expression).getDynamicConjuncts().stream())
+                        .filter(descriptor -> descriptor.getId().equals(dynamicFilterId))
+                        .map(descriptor -> Symbol.from(descriptor.getInput()))
+                        .allMatch(sourceSymbol -> symbolAliases.get(sourceSymbolAlias).equals(sourceSymbol.toSymbolReference()));
```
you should make sure that there is at least one match (`allMatch` returns true on empty stream), e.g:
```
List<DynamicFilter.Descriptor> matchingDescriptors = searchFrom(semiJoinNode.getSource())
  .where(FilterNode.class::isInstance)
  .findAll()
  .stream()
  .flatMap(filterNode -> extractExpressions(filterNode).stream())
  .flatMap(expression -> extractDynamicFilters(expression).getDynamicConjuncts().stream())
  .filter(descriptor -> descriptor.getId().equals(dynamicFilterId))
  .collect(toImmutableList();
boolean sourceSymbolsMatch = matchingDescriptors.stream()
  .map(descriptor -> Symbol.from(descriptor.getInput()))
  .allMatch(sourceSymbol -> symbolAliases.get(sourceSymbolAlias).equals(sourceSymbol.toSymbolReference()));
if (!matchingDescriptors.isEmpty() && sourceSymbolsMatch) {
  ...
}
```

lxynov replied:
oh good catch. Thanks for the pointer

sopel39 reviewed (approved):
small comments

sopel39 commented:
merged, thanks!
